tests/vmipreset_test.go

```diff
@@ -297,21 +297,59 @@ var _ = Describe("[rfe_id:609][crit:medium][vendor:cnv-qe@redhat.com][level:comp
 		It("[test_id:1605]should denied to start the VMI", func() {
 			err := virtClient.RestClient().Post().Resource("virtualmachineinstancepresets").Namespace(tests.NamespaceTestDefault).Body(conflictPreset).Do().Error()
 			Expect(err).ToNot(HaveOccurred())
-			waitForPreset(virtClient, conflictPrefix)
+			waitForPreset(virtClient, conflictPrefix, true)
 
 			err = virtClient.RestClient().Post().Resource("virtualmachineinstancepresets").Namespace(tests.NamespaceTestDefault).Body(memoryPreset).Do().Error()
 			Expect(err).ToNot(HaveOccurred())
-			waitForPreset(virtClient, memoryPrefix)
+			waitForPreset(virtClient, memoryPrefix, true)
 
 			vmi.Labels = map[string]string{flavorKey: memoryFlavor, conflictKey: conflictFlavor}
 			By("creating the VirtualMachineInstance")
 			_, err = virtClient.VirtualMachineInstance(tests.NamespaceTestDefault).Create(vmi)
 			Expect(err).To(HaveOccurred())
 		})
 	})
+
+	Context("Preset Lifecycle", func() {
+		var preset *v1.VirtualMachineInstancePreset
+		presetNamePrefix := "vmi-preset-small-"
+		selectorKey := "kubevirt.io/vmPreset"
+		selectorLabel := "vmi-preset-small"
+
+		BeforeEach(func() {
+			selector := k8smetav1.LabelSelector{MatchLabels: map[string]string{selectorKey: selectorLabel}}
+			memory, _ := resource.ParseQuantity("64M")
+			preset = &v1.VirtualMachineInstancePreset{
+				ObjectMeta: k8smetav1.ObjectMeta{GenerateName: presetNamePrefix},
+				Spec: v1.VirtualMachineInstancePresetSpec{
+					Selector: selector,
+					Domain: &v1.DomainSpec{
+						Resources: v1.ResourceRequirements{Requests: k8sv1.ResourceList{
+							"memory": memory}},
+					},
+				},
+			}
+		})
+
+                It("[test_id:617][rfe_id:609] should create and delete preset", func() {
+			By("Creating preset")
+			err := virtClient.RestClient().Post().Resource("virtualmachineinstancepresets").Namespace(tests.NamespaceTestDefault).Body(preset).Do().Error()
+			Expect(err).ToNot(HaveOccurred())
+
+			By("Checking that preset was created")
+			newPreset := waitForPreset(virtClient, presetNamePrefix, true)
```

Comment:
I don't understand this waitForPreset() function. The preset should exist right after it's posted right? What is being waited on?

Reply:
etcd is a distributed keystore is it not? Why assume posts are instantaneous? The normal codepath here would just be a get anyways.

Reply:
> Why assume posts are instantaneous?

This is inaccurate and needs to be cleared up.

A successful POST is not eventually consistent like this code implies. When you make a successful POST, a GET on the same object will return the latest object. There's never a reason to "wait" for an object to be available once a successful POST is issued.

Reply:
@davidvossel, please read a couple of links:

https://kubernetes.io/docs/tasks/administer-cluster/highly-available-master/

https://raft.github.io/

You can skim them, that's fine. The point is your assertion is not categorically true. In the simple case, it's *usually* true that a POST means a GET will return the correct data, but it's not *always* true. In particular, in a multi-master cluster with `ENABLE_ETCD_QUORUM_READ=false`, it's possible to have an eventually-consistent model. I personally dislike the idea of purposely writing less robust code, but if you insist, then I'll change it in the interests of just moving on.

Reply:
> the point is your assertion is not categorically true. 

umm... i really don't think i'm wrong here. this appears true since k8s 1.9. 

https://github.com/kubernetes/kubernetes/pull/53795
https://github.com/kubernetes/kubernetes/pull/53875

That k8s documentation you pulled up is 5 years old. 

> personally dislike the idea of purposely writing less robust code, but if you insist, then I'll change it in the interests of just moving on.

forgetting eventual consistency for a second, what's going on here isn't robust regardless.

Guaranteeing that you can GET a VMPreset value from the apiserver in a Functional Test does not guarantee that virt-api's informers have populated the value in the mutation webhook... That's really what you're wanting here. It just so happens that performing the GET request introduces enough delay for virt-api's informers to populate the preset in cache before the VMI is posted. 

So, If you post a VMPreset and immediately post a VMI, then it's possible that the VMI doesn't have the VMPreset applied... This isn't because GETS are eventually consistent, this is because virt-api is a separate process that has to be informed of the VMPreset. It's a race condition between two processes. Performing the GET guarantees nothing other than a little more time is consumed before the VMI is posted which makes it look like it's helping somehow.

Reply:
Ok so back to what you said at first. If a consecutive POST and GET of a resource is always guaranteed to work, then none of this matters. The tests that need to fetch a new copy for validation still need to do that--so the intrinsic delay in the client side behavior will always be there, and will in fact be effectively identical whether we do a GET directly or a GET inside an eventually clause (the first GET is guaranteed to work anyways).

Since these tests have never hit our radar as notoriously flakey, I don't expect anything will change if we just remove that code.

Reply:
> The tests that need to fetch a new copy for validation still need to do that

FYI, You can retrieve the resulting preset that was persisted into etd as the result of the initial Create() POST. Below is an example.  

```
newPreset, err := virtClient.VirtualMachineInstance(tests.NamespaceTestDefault).Create(preset)
// sleep allows virt-api's informer to cache preset before we post a VMI using the preset. 
time.Sleep(time.Second)
```

> so the intrinsic delay in the client side behavior will always be there, and will in fact be effectively identical whether we do a GET directly or a GET inside an eventually clause (the first GET is guaranteed to work anyways).

Right, the tests pass because of the timing delay the subsequent GET introduces after the object is POSTed.

So yes, the tests pass. But it's kind of by luck rather being intentional about accounting for the virt-api timing gap. I'd rather we explicitly accounted for that delay so it's obvious to everyone that the delay is what's holding the test together. 

We can follow up on that later if you'd like.

Reply:
I agree that it's worth investigating. I also agree that it's beyond the scope of this PR. Let's keep that as a separate effort